Add checks for structured in-place operations.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65686

Fixes: #57827

This PR introduces `check_inplace` function. It contains some common checks for all
structured in-place operators (e.g. dtype, device, and sizes). `set_output` method calls
`check_inplace` on in-place specializations of structured kernels.

Besides that, it also:
- adds overlap assertions for both in-place and out-of-place overloads
- remove in-place operator specific `TORCH_CHECK` around the code base

Differential Revision: [D31234063](https://our.internmc.facebook.com/intern/diff/D31234063)

ysiraichi committed:
Add checks for structured in-place operations.

Fixes: #57827

This PR introduces `check_inplace` function. It contains some common checks for all
structured in-place operators (e.g. dtype, device, and sizes). `set_output` method calls
`check_inplace` on in-place specializations of structured kernels. Finally, it also adds
overlap assertions for both in-place and out-of-place overloads.

[ghstack-poisoned]

ysiraichi committed:
Update on "Add checks for structured in-place operations."


Fixes: #57827

This PR introduces `check_inplace` function. It contains some common checks for all
structured in-place operators (e.g. dtype, device, and sizes). `set_output` method calls
`check_inplace` on in-place specializations of structured kernels. Finally, it also adds
overlap assertions for both in-place and out-of-place overloads.

[ghstack-poisoned]

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -84,15 +84,34 @@ def gen_resize_out_helper(backend_index: BackendIndex) -> List[str]:
       out.unsafeGetTensorImpl()->empty_tensor_restride(*options.memory_format_opt());
     }
+  } else {
+    at::assert_no_internal_overlap(out);
```
This is failing you in tests and you're not allowed to do it anyway, because some operators are OK with internal overlap and structured doesn't know which ones are which.

ezyang commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -84,15 +84,34 @@ def gen_resize_out_helper(backend_index: BackendIndex) -> List[str]:
+      "Bad in-place call: ",
+      "input tensor size ", self.sizes(), " and output tensor size ", sizes, " should match");
+  at::assert_no_internal_overlap(self);
+}
```
There's a subtlety here which is probably worth adding comments for, which is that some of the tests that you see here are done separately for TensorIterator kernels, inside TensorIterator. So while this is required for things like addmm (which don't do TensorIterator), it is redundant for other operators (like add) and are a slight regression in instruction count. I don't think this PR should get blocked because of this, especially because ufunc codegen is coming along well and should make it easier to fix this problem, it's just something to be aware of.

ezyang commented:
LGTM, phone me up when all the tests pass.

ysiraichi committed:
Update on "Add checks for structured in-place operations."


Fixes: #57827

This PR introduces `check_inplace` function. It contains some common checks for all
structured in-place operators (e.g. dtype, device, and sizes). `set_output` method calls
`check_inplace` on in-place specializations of structured kernels.

Besides that, it also:
- adds overlap assertions for both in-place and out-of-place overloads
- remove in-place operator specific `TORCH_CHECK` around the code base

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31234063).
